io/genbank/parser.go

```diff
@@ -0,0 +1,356 @@
+	if err != nil {
+		return err
+	}
+
+	for prefix, parseFunc := range funcs {
+		if strings.HasPrefix(line, prefix) {
+			return parseFunc(ctx)
+		}
+	}
+
+	p.readLine() // We only peeked previously, so we have to read for proper error reporting
+	return p.makeSyntaxError("keyword expected, none found")
+}
+
+/* MAIN PARSER FUNCTIONS */
+
+// Parse parses a Genbank file from a Parser's
+// io.Reader.
+func (p *Parser) Parse() (Genbank, error) {
+	res := Genbank{}
+
+	header, err := p.parseHeader()
+	if err != nil {
+		return res, fmt.Errorf("failed to parse header: %w", err)
+	}
+	res.Header = header
+
+	res.Entries = make([]Entry, 0)
+	var entry Entry
+	reachedEOF := false
+	for !reachedEOF {
+		entry, reachedEOF, err = p.parseEntry()
+		if err != nil {
+			return res, fmt.Errorf("failed to parse entry %v: %w", len(res.Entries), err)
+		}
+		res.Entries = append(res.Entries, entry)
+
+	}
+
+	return res, nil
+}
+
+const headerDateLayout = "January 2 2006"
+
+func (p *Parser) parseHeader() (Header, error) {
+	res := Header{}
+
+	fileNameLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read file name: %w", err)
+	}
+	fileName := strings.TrimSpace(fileNameLine[:10])
+	if fileName == "" {
+		return res, p.makeSyntaxError("empty file name")
+	}
+	res.FileName = fileName
+
+	dateLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read date: %w", err)
+	}
+	date, err := time.Parse(headerDateLayout, strings.TrimSpace(dateLine))
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse date", err)
+	}
+	res.Date = date
+
+	p.readLine() // Skip empty line
+
+	releaseLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read release: %w", err)
+	}
+	if len(releaseLine) < 48 {
+		return res, p.makeSyntaxError("release line too short")
+	}
+	release := strings.Split(strings.TrimSpace(releaseLine[47:]), ".")
+	if len(release) != 2 {
+		return res, p.makeSyntaxError("malformed release version")
+	}
+	res.MajorRelease, err = strconv.Atoi(release[0])
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse major release", err)
+	}
+	res.MinorRelease, err = strconv.Atoi(release[1])
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse minor release", err)
+	}
+
+	p.readLine() // Skip empty line
+
+	titleLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read title: %w", err)
+	}
+	res.Title = strings.TrimSpace(titleLine)
+
+	p.readLine() // Skip empty line
+
+	statsLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read file statistics: %w", err)
+	}
+	if len(releaseLine) < 48 {
+		return res, p.makeSyntaxError("release line too short")
+	}
+	res.NumEntries, err = strconv.Atoi(strings.TrimSpace(statsLine[:8]))
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse number of entries/loci", err)
+	}
+	res.NumBases, err = strconv.Atoi(strings.TrimSpace(statsLine[15:26]))
```

Comment:
The spec has a lot of opinions about spacing. I'm not sure what the best way to handle this is. Perhaps this kind of thing is the best approach. Maybe it could help to abstract this a bit? This could help with reading why the numbers are there, but more importantly perhaps it can also be used for writing? 

For instance, maybe something like 

type TokenRange struct {
  start int
  end int
}

var fileNameRange = TokenRange{15,26}
var statsRange =  TokenRange{39,47}

that way slicing could look like

statsLine[statsRange.start:statsRange.end]

Maybe it is helpful for formatting text for writing as well? 

Just spit balling.

Reply:
Absolutely! Good point re: being helpful for writing. Will go back and do this.

Reply:
Done now. I think I got all (or most of) the magic numbers. Proved useful with the header write ability!